zcbor/zcbor.py

```diff
@@ -2448,12 +2448,20 @@ def result_len(self):
     # key, cbor, and repetitions.
     def full_xcode(self, union_int=None):
         if self.present_var_condition():
-            func, *arguments = self.repeated_single_func(ptr_result=True)
-            return (
-                f"zcbor_present_{self.mode}(&(%s), (zcbor_{self.mode}r_t *)%s, %s)" %
-                (self.present_var_access(),
-                 func,
-                 xcode_args(*arguments),))
+            if self.mode == "encode":
+                func, *arguments = self.repeated_single_func(ptr_result=False)
+                return f"(!{self.present_var_access()} || {func}({xcode_args(*arguments)}))"
+            else:
+                if not self.repeated_single_func_impl_condition():
+                    decode_str = self.repeated_xcode(union_int)
+                    assert "zcbor_" in decode_str, \
+                        """Must be a direct call to zcbor to guarantee that payload and elem_count"
+are cleaned up after a failure."""
+                    return f"({self.present_var_access()} = {self.repeated_xcode(union_int)}, 1)"
+                func, *arguments = self.repeated_single_func(ptr_result=True)
+                return (
+                    f"zcbor_present_{self.mode}(&(%s), (zcbor_{self.mode}r_t *)%s, %s)" %
```

Comment:
Can't this end up generating calls to `zcbor_present_encode`?

Reply:
It's under an `else` clause that excludes self.mode being "encode" but I agree it's a bit confusing, I changed it to literal "decode".